File: backend/app/metrics.py
"""进程内指标（Prometheus 文本格式暴露，零第三方依赖）。

计数器带标签（如 agent/status）；直方图为固定桶（毫秒）。
线程安全：asyncio 事件循环与 asyncio.to_thread 两侧调用均加锁。
"""
from __future__ import annotations

import threading
import time
from collections import defaultdict

DEFAULT_BUCKETS_MS = (1_000, 5_000, 10_000, 30_000, 60_000, 120_000, 300_000)


def _render_labels(labels: dict[str, str]) -> str:
    if not labels:
        return ""
    inner = ",".join(f'{k}="{v}"' for k, v in sorted(labels.items()))
    return "{" + inner + "}"


class Metrics:
    def __init__(self, buckets_ms: tuple[int, ...] = DEFAULT_BUCKETS_MS):
        self._lock = threading.Lock()
        self._counters: dict[str, dict[tuple, int]] = defaultdict(lambda: defaultdict(int))
        self._histograms: dict[str, dict[tuple, dict]] = defaultdict(dict)
        self._buckets = tuple(sorted(buckets_ms))
        self._process_start = time.time()

    def inc_counter(self, name: str, value: int = 1, **labels: str) -> None:
        key = tuple(sorted(labels.items()))
        with self._lock:
            self._counters[name][key] += value

    def observe(self, name: str, value_ms: float) -> None:
        with self._lock:
            bucket_state = self._histograms[name].setdefault(
                (), {"counts": [0] * (len(self._buckets) + 1), "sum": 0.0, "count": 0}
            )
            idx = len(self._buckets)
            for i, bound in enumerate(self._buckets):
                if value_ms <= bound:
                    idx = i
                    break
            bucket_state["counts"][idx] += 1
            bucket_state["sum"] += value_ms
            bucket_state["count"] += 1

    # --- 暴露 ---

    def render(self) -> str:
        with self._lock:
            lines: list[str] = []
            uptime = time.time() - self._process_start
            lines.append("# HELP piwren_process_uptime_seconds Process uptime in seconds")
            lines.append("# TYPE piwren_process_uptime_seconds gauge")
            lines.append(f"piwren_process_uptime_seconds {uptime:.1f}")

            for name, series in sorted(self._counters.items()):
                metric = f"piwren_{name}_total"
                lines.append(f"# HELP {metric} Counter {name}")
                lines.append(f"# TYPE {metric} counter")
                for key, value in sorted(series.items()):
                    lines.append(f"{metric}{_render_labels(dict(key))} {value}")

            for name, _ in sorted(self._histograms.items()):
                metric = f"piwren_{name}_milliseconds"
                lines.append(f"# HELP {metric} Histogram {name} in milliseconds")
                lines.append(f"# TYPE {metric} histogram")
                state = self._histograms[name].get(())
                if not state:
                    continue
                cumulative = 0
                for bound, cnt in zip(self._buckets, state["counts"]):
                    cumulative += cnt
                    lines.append(f'{metric}_bucket{{le="{bound}"}} {cumulative}')
                cumulative += state["counts"][-1]
                lines.append(f'{metric}_bucket{{le="+Inf"}} {cumulative}')
                lines.append(f'{metric}_sum {state["sum"]:.1f}')
                lines.append(f'{metric}_count {state["count"]}')
            return "\n".join(lines) + "\n"


metrics = Metrics()

File: services/pi-orchestrator/src/metrics.mjs
/** 进程内指标（Prometheus 文本格式暴露，零依赖；与后端 app/metrics.py 同思路）。
 *
 * - 计数器：incCounter(name, labels?, value?)，暴露为 piwrenpi_<name>_total
 * - 直方图：observe(name, valueMs, labels?)，固定桶（毫秒），暴露为 piwrenpi_<name>_milliseconds
 * 线程模型：Node 单线程事件循环，无需加锁。
 */

const BUCKETS_MS = [500, 1_000, 5_000, 10_000, 30_000, 60_000, 120_000, 180_000];

const labelKey = (labels) =>
  Object.keys(labels ?? {})
    .sort()
    .map((k) => `${k}="${String(labels[k]).replace(/\\/g, '\\\\').replace(/"/g, '\\"')}"`)
    .join(',');

const renderLabels = (key) => (key ? `{${key}}` : '');

/** 桶线标签：数据标签 + le 合并进同一个 {} 块（Prometheus 要求单标签块）。 */
const renderBucketLabels = (key, le) => (key ? `{${key},${le}}` : `{${le}}`);

export function createMetrics({ bucketsMs = BUCKETS_MS } = {}) {
  const buckets = [...bucketsMs].sort((a, b) => a - b);
  const counters = new Map(); // name -> Map(labelKey -> value)
  const histograms = new Map(); // name -> Map(labelKey -> { counts, sum, count })

  const seriesOf = (map, name) => {
    let series = map.get(name);
    if (!series) {
      series = new Map();
      map.set(name, series);
    }
    return series;
  };

  return {
    incCounter(name, labels = undefined, value = 1) {
      const series = seriesOf(counters, name);
      const key = labelKey(labels);
      series.set(key, (series.get(key) ?? 0) + value);
    },

    observe(name, valueMs, labels = undefined) {
      const series = seriesOf(histograms, name);
      const key = labelKey(labels);
      let state = series.get(key);
      if (!state) {
        state = { counts: new Array(buckets.length + 1).fill(0), sum: 0, count: 0 };
        series.set(key, state);
      }
      let idx = buckets.length;
      for (let i = 0; i < buckets.length; i += 1) {
        if (valueMs <= buckets[i]) {
          idx = i;
          break;
        }
      }
      state.counts[idx] += 1;
      state.sum += valueMs;
      state.count += 1;
    },

    render() {
      const lines = [];
      lines.push('# HELP piwrenpi_process_uptime_seconds Process uptime in seconds');
      lines.push('# TYPE piwrenpi_process_uptime_seconds gauge');
      lines.push(`piwrenpi_process_uptime_seconds ${process.uptime().toFixed(1)}`);

      for (const name of [...counters.keys()].sort()) {
        const metric = `piwrenpi_${name}_total`;
        lines.push(`# HELP ${metric} Counter ${name}`);
        lines.push(`# TYPE ${metric} counter`);
        const series = counters.get(name);
        for (const key of [...series.keys()].sort()) {
          lines.push(`${metric}${renderLabels(key)} ${series.get(key)}`);
        }
      }

      for (const name of [...histograms.keys()].sort()) {
        const metric = `piwrenpi_${name}_milliseconds`;
        lines.push(`# HELP ${metric} Histogram ${name} in milliseconds`);
        lines.push(`# TYPE ${metric} histogram`);
        const series = histograms.get(name);
        for (const key of [...series.keys()].sort()) {
          const state = series.get(key);
          let cumulative = 0;
          buckets.forEach((bound, i) => {
            cumulative += state.counts[i];
            lines.push(`${metric}_bucket${renderBucketLabels(key, `le="${bound}"`)} ${cumulative}`);
          });
          cumulative += state.counts[buckets.length];
          lines.push(`${metric}_bucket${renderBucketLabels(key, 'le="+Inf"')} ${cumulative}`);
          lines.push(`${metric}_sum${renderLabels(key)} ${state.sum.toFixed(1)}`);
          lines.push(`${metric}_count${renderLabels(key)} ${state.count}`);
        }
      }
      return `${lines.join('\n')}\n`;
    },
  };
}
